diff --git a/Modules.make b/Modules.make
index 765fd38..7a5b1d1 100644
--- a/Modules.make
+++ b/Modules.make
@@ -25,7 +25,7 @@ MODULES += MODULE_ADAPT_MLLR
 MODULES += MODULE_ADAPT_ADVANCED
 
 # ****** Audio ******
-MODULES += MODULE_AUDIO_FFMPEG
+# MODULES += MODULE_AUDIO_FFMPEG
 MODULES += MODULE_AUDIO_FLAC
 MODULES += MODULE_AUDIO_HTK
 MODULES += MODULE_AUDIO_OSS
@@ -33,7 +33,7 @@ MODULES += MODULE_AUDIO_RAW
 MODULES += MODULE_AUDIO_WAV_SYSTEM
 
 # ****** Cache Manager integration ******
-# MODULES += MODULE_CORE_CACHE_MANAGER
+MODULES += MODULE_CORE_CACHE_MANAGER
 
 # ****** Cart ******
 MODULES += MODULE_CART
@@ -69,11 +69,11 @@ MODULES += MODULE_THEANO_INTERFACE
 MODULES += MODULE_PYTHON
 
 # ****** OpenFst ******
-MODULES += MODULE_OPENFST
+# MODULES += MODULE_OPENFST
 
 # ****** Search ******
 MODULES += MODULE_SEARCH_MBR
-MODULES += MODULE_SEARCH_WFST
+# MODULES += MODULE_SEARCH_WFST
 MODULES += MODULE_SEARCH_LINEAR
 MODULES += MODULE_ADVANCED_TREE_SEARCH
 MODULES += MODULE_GENERIC_SEQ2SEQ_TREE_SEARCH
diff --git a/config/os-linux.make b/config/os-linux.make
index bad41bf..eaa0fb4 100644
--- a/config/os-linux.make
+++ b/config/os-linux.make
@@ -46,18 +46,12 @@ LDFLAGS     += -L$(TBB_DIR)/lib -ltbb
 endif
 
 ifdef MODULE_TENSORFLOW
-TF_COMPILE_BASE = /opt/tensorflow/tensorflow
-
 TF_CXXFLAGS  = -fexceptions
-TF_CXXFLAGS += -I$(TF_COMPILE_BASE)/
-TF_CXXFLAGS += -I$(TF_COMPILE_BASE)/bazel-bin/
-TF_CXXFLAGS += -I$(TF_COMPILE_BASE)/bazel-genfiles/
-TF_CXXFLAGS += -I$(TF_COMPILE_BASE)/bazel-tensorflow/external/eigen_archive/
-TF_CXXFLAGS += -I$(TF_COMPILE_BASE)/bazel-tensorflow/external/com_google_protobuf/src/
-TF_CXXFLAGS += -I$(TF_COMPILE_BASE)/bazel-tensorflow/external/com_google_absl/
-
-TF_LDFLAGS  = -L$(TF_COMPILE_BASE)/bazel-bin/tensorflow -ltensorflow_cc -ltensorflow_framework
-TF_LDFLAGS += -Wl,-rpath -Wl,$(TF_COMPILE_BASE)/bazel-bin/tensorflow
+TF_CXXFLAGS += -I/usr/local/lib/python3.8/dist-packages/tensorflow/include
+TF_LDFLAGS   = /usr/local/lib/tensorflow/libtensorflow_cc.so.2
+TF_LDFLAGS  += /usr/local/lib/python3.8/dist-packages/tensorflow/libtensorflow_framework.so.2
+TF_LDFLAGS  += -Wl,--allow-multiple-definition -Wl,--whole-archive -Wl,--no-as-needed
+TF_LDFLAGS  += -Wl,-rpath -Wl,/usr/local/lib/tensorflow
 
 # USE_TENSORFLOW_MKL=1
 endif
@@ -109,16 +103,15 @@ endif
 endif
 
 ifdef MODULE_CUDA
-CUDAROOT    = /usr/local/cuda-7.0
+CUDAROOT    = /usr/local/cuda-11.6
 INCLUDES    += -I$(CUDAROOT)/include/
 LDFLAGS     += -L$(CUDAROOT)/lib64/ -lcublas -lcudart -lcurand
 NVCC        = $(CUDAROOT)/bin/nvcc
 # optimal for GTX680; set sm_35 for K20
-NVCCFLAGS   = -gencode arch=compute_20,code=sm_20 \
-	      -gencode arch=compute_30,code=sm_30 \
-	      -gencode arch=compute_35,code=sm_35 \
-	      -gencode arch=compute_52,code=sm_52 \
-	      -gencode arch=compute_61,code=sm_61
+NVCCFLAGS   = -gencode arch=compute_61,code=sm_61 \  # GTX 1080
+	      -gencode arch=compute_75,code=sm_75 \  # RTX 2080
+	      -gencode arch=compute_86,code=sm_86 \  # RTX 3090
+	      --compiler-options -fPIC
 endif
 
 ifeq ($(PROFILE),gprof)
@@ -148,15 +141,16 @@ ifdef MODULE_PYTHON
 PYTHON_PATH =
 ifneq (${PYTHON_PATH},)
 INCLUDES    += `${PYTHON_PATH}/bin/python3-config --includes 2>/dev/null`
-LDFLAGS     += `${PYTHON_PATH}/bin/python3-config --ldflags 2>/dev/null`
+LDFLAGS     += `${PYTHON_PATH}/bin/python3-config --ldflags --embed 2>/dev/null`
 LDFLAGS     += -Wl,-rpath -Wl,${PYTHON_PATH}/lib
 else
 INCLUDES    += `python3-config --includes 2>/dev/null`
-LDFLAGS     += `python3-config --ldflags 2>/dev/null`
+LDFLAGS     += `python3-config --ldflags --embed 2>/dev/null`
 # IF you want to use Python2 for whatever reason:
 # INCLUDES    += `pkg-config --cflags python`
 # LDFLAGS     += `pkg-config --libs python`
 endif
+INCLUDES    += -I$(shell python3 -c 'import numpy as np; print(np.get_include())')
 endif
 
 # X11 and QT
